// be-orgchart/esm2015/lib/components/orgchart/be-orgchart.component.js
import { __decorate } from "tslib";
import { Component, ElementRef, AfterViewInit, Input, Output, OnChanges, SimpleChanges, EventEmitter } from '@angular/core';
import { D3OrgChart } from '../../clasess/d3-org-chart.class';
let BEOrgchartComponent = class BEOrgchartComponent {
    //#endregion
    constructor(prEl) {
        this.prEl = prEl;
        this.onNodeClick = new EventEmitter();
    }
    get chart() {
        const me = this;
        if (!me._chart) {
            const options = {};
            // adding nodes if defined
            if (me.nodes)
                options.data = me.nodes;
            // adding node parser if defined
            if (me.nodeParser)
                options.nodeParser = me.nodeParser;
            me._chart = new D3OrgChart(this.prEl.nativeElement, options);
        }
        ;
        return me._chart;
    }
    ngAfterViewInit() {
        this.init();
    }
    ngOnChanges(changes) {
        const me = this;
        // detecting changes of input
        if (changes.nodeParser && changes.nodeParser.currentValue) {
            me.assignNodeParser(changes.nodeParser.currentValue);
            me.chart.render();
        }
        if (changes.nodes && changes.nodes.currentValue) {
            me.assignData(changes.nodes.currentValue);
            me.chart.render();
        }
    }
    init() {
        const me = this;
        // init the canvas
        const chart = me.chart; // Must be like this to ensure chart init
        setTimeout(() => me.chart.render(), 200);
        // setting data
        me.chart.onNodeClick.subscribe((data) => {
            me.onNodeClick.next(data.node);
        });
    }
    assignData(data, clear = true) {
        const me = this;
        const chart = me.chart; // Must be like this to ensure chart init
        // if isn´t array we convert it in array
        if (!(data instanceof Array))
            data = [data];
        if (clear) {
            chart.data = data;
            return;
        }
        // pushing data
        chart.data = [...chart.data, ...data];
        chart.render();
    }
    assignNodeParser(prParser) {
        const me = this;
        const chart = me.chart; // Must be like this to ensure chart init
        chart.nodeParser = prParser;
        chart.render();
    }
    assignOptions(prOptions) {
        const me = this;
        me.chart.setOptions(prOptions);
    }
};
BEOrgchartComponent.ctorParameters = () => [
    { type: ElementRef }
];
__decorate([
    Input()
], BEOrgchartComponent.prototype, "nodes", void 0);
__decorate([
    Input()
], BEOrgchartComponent.prototype, "nodeParser", void 0);
__decorate([
    Input()
], BEOrgchartComponent.prototype, "options", void 0);
__decorate([
    Output()
], BEOrgchartComponent.prototype, "onNodeClick", void 0);
BEOrgchartComponent = __decorate([
    Component({
        selector: 'be-orgchart',
        template: "<div #orgchart class=\"container\"></div>",
        styles: [":host{display:flex;flex-direction:column;overflow:hidden}"]
    })
], BEOrgchartComponent);
export { BEOrgchartComponent };
