// ngx-org-chart/__ivy_ngcc__/fesm2015/ngx-org-chart.js
import { CommonModule } from '@angular/common';
import { EventEmitter, Component, Input, Output, NgModule, HostBinding } from '@angular/core';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';

function NgxChartNodeComponent_div_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "div", 5);
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵpropertyInterpolate1("ngClass", "ngx-org-connector-", ctx_r0.direction, "");
} }
const _c0 = function (a0) { return { "background-image": a0 }; };
function NgxChartNodeComponent_div_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "div", 6);
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngStyle", ɵngcc0.ɵɵpureFunction1(1, _c0, "url('" + (ctx_r1.node == null ? null : ctx_r1.node.image) + "')"));
} }
function NgxChartNodeComponent_div_8_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "div", 5);
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵpropertyInterpolate1("ngClass", "ngx-org-connector-", ctx_r2.direction, "");
} }
function NgxChartDesignerComponent_ngx_chart_node_0_Template(rf, ctx) { if (rf & 1) {
    const _r3 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "ngx-chart-node", 2);
    ɵngcc0.ɵɵlistener("itemClick", function NgxChartDesignerComponent_ngx_chart_node_0_Template_ngx_chart_node_itemClick_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r3); const ctx_r2 = ɵngcc0.ɵɵnextContext(); return ctx_r2.itemClick.emit($event); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵpropertyInterpolate1("ngClass", "ngx-org-", ctx_r0.direction, "");
    ɵngcc0.ɵɵproperty("node", ctx_r0.node)("hasParent", ctx_r0.hasParent)("direction", ctx_r0.direction);
} }
function NgxChartDesignerComponent_div_1_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    const _r9 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "div", 5);
    ɵngcc0.ɵɵelementStart(2, "div", 6);
    ɵngcc0.ɵɵelement(3, "div", 7);
    ɵngcc0.ɵɵelement(4, "div", 8);
    ɵngcc0.ɵɵelement(5, "div", 7);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(6, "ngx-chart-designer", 9);
    ɵngcc0.ɵɵlistener("itemClick", function NgxChartDesignerComponent_div_1_ng_container_1_Template_ngx_chart_designer_itemClick_6_listener($event) { ɵngcc0.ɵɵrestoreView(_r9); const ctx_r8 = ɵngcc0.ɵɵnextContext(2); return ctx_r8.itemClick.emit($event); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const childNode_r5 = ctx.$implicit;
    const isFirst_r6 = ctx.first;
    const isLast_r7 = ctx.last;
    const ctx_r4 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate1("ngClass", "ngx-org-org-container-", ctx_r4.direction, "");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate1("ngClass", "ngx-org-connector-container-", ctx_r4.direction, "");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵstyleProp("border-color", isFirst_r6 ? "transparent" : "");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵstyleProp("border-color", isLast_r7 ? "transparent" : "");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("node", childNode_r5)("hasParent", true)("direction", ctx_r4.direction);
} }
function NgxChartDesignerComponent_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 3);
    ɵngcc0.ɵɵtemplate(1, NgxChartDesignerComponent_div_1_ng_container_1_Template, 7, 9, "ng-container", 4);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵpropertyInterpolate1("ngClass", "ngx-org-reports-", ctx_r1.direction, "");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r1.node == null ? null : ctx_r1.node.childs);
} }
function NgxOrgChartComponent_div_0_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    const _r4 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "div");
    ɵngcc0.ɵɵelementStart(2, "ngx-chart-designer", 2);
    ɵngcc0.ɵɵlistener("itemClick", function NgxOrgChartComponent_div_0_ng_container_1_Template_ngx_chart_designer_itemClick_2_listener($event) { ɵngcc0.ɵɵrestoreView(_r4); const ctx_r3 = ɵngcc0.ɵɵnextContext(2); return ctx_r3.itemClick.emit($event); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const node_r2 = ctx.$implicit;
    const ctx_r1 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵclassMapInterpolate1("ngx-org-self-", ctx_r1.direction, "");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("node", node_r2)("direction", ctx_r1.direction);
} }
function NgxOrgChartComponent_div_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div");
    ɵngcc0.ɵɵtemplate(1, NgxOrgChartComponent_div_0_ng_container_1_Template, 3, 5, "ng-container", 1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r0.nodes);
} }
class NgxChartNodeComponent {
    constructor() {
        this.hasParent = false;
        this.direction = 'vertical';
        this.itemClick = new EventEmitter();
    }
}
NgxChartNodeComponent.ɵfac = function NgxChartNodeComponent_Factory(t) { return new (t || NgxChartNodeComponent)(); };
NgxChartNodeComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: NgxChartNodeComponent, selectors: [["ngx-chart-node"]], inputs: { hasParent: "hasParent", direction: "direction", node: "node" }, outputs: { itemClick: "itemClick" }, decls: 9, vars: 6, consts: [["class", "ngx-org-border", 3, "ngClass", 4, "ngIf"], [1, "ngx-org-box", "ngx-org-border", "ngx-org-background", 3, "ngClass", "click"], ["class", "ngx-org-image ngx-org-border", 3, "ngStyle", 4, "ngIf"], [1, "ngx-org-name"], [1, "ngx-org-title"], [1, "ngx-org-border", 3, "ngClass"], [1, "ngx-org-image", "ngx-org-border", 3, "ngStyle"]], template: function NgxChartNodeComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, NgxChartNodeComponent_div_0_Template, 1, 1, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵlistener("click", function NgxChartNodeComponent_Template_div_click_1_listener() { return ctx.itemClick.emit(ctx.node); });
        ɵngcc0.ɵɵtemplate(2, NgxChartNodeComponent_div_2_Template, 1, 3, "div", 2);
        ɵngcc0.ɵɵelementStart(3, "div");
        ɵngcc0.ɵɵelementStart(4, "div", 3);
        ɵngcc0.ɵɵtext(5);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(6, "div", 4);
        ɵngcc0.ɵɵtext(7);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(8, NgxChartNodeComponent_div_8_Template, 1, 1, "div", 0);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", ctx.hasParent);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngClass", ctx.node == null ? null : ctx.node.cssClass);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.node == null ? null : ctx.node.image);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵtextInterpolate(ctx.node == null ? null : ctx.node.name);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate(ctx.node == null ? null : ctx.node.title);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.node == null ? null : ctx.node.childs == null ? null : ctx.node.childs.length);
    } }, directives: [ɵngcc1.NgIf, ɵngcc1.NgClass, ɵngcc1.NgStyle], styles: ["[_nghost-%COMP%]{display:flex;align-items:center}.ngx-org-box[_ngcontent-%COMP%]{cursor:pointer;display:flex;align-items:center}"] });
NgxChartNodeComponent.propDecorators = {
    node: [{ type: Input }],
    hasParent: [{ type: Input }],
    direction: [{ type: Input }],
    itemClick: [{ type: Output }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(NgxChartNodeComponent, [{
        type: Component,
        args: [{
                selector: 'ngx-chart-node',
                template: "<div ngClass=\"ngx-org-connector-{{direction}}\" class=\"ngx-org-border\" *ngIf=\"hasParent\"></div>\r\n<div [ngClass]=\"node?.cssClass\" class=\"ngx-org-box ngx-org-border ngx-org-background\" (click)=\"itemClick.emit(node)\">\r\n  <div *ngIf=\"node?.image\" class=\"ngx-org-image ngx-org-border\"\r\n    [ngStyle]=\"{ 'background-image': 'url(\\''+ node?.image+'\\')'}\">\r\n  </div>\r\n  <div>\r\n    <div class=\"ngx-org-name\">{{node?.name}}</div>\r\n    <div class=\"ngx-org-title\">{{node?.title}}</div>\r\n  </div>\r\n</div>\r\n<div ngClass=\"ngx-org-connector-{{direction}}\" class=\"ngx-org-border\" *ngIf=\"node?.childs?.length\"></div>\r\n",
                styles: [":host{display:flex;align-items:center}.ngx-org-box{cursor:pointer;display:flex;align-items:center}"]
            }]
    }], function () { return []; }, { hasParent: [{
            type: Input
        }], direction: [{
            type: Input
        }], itemClick: [{
            type: Output
        }], node: [{
            type: Input
        }] }); })();
if (false) {
    /** @type {?} */
    NgxChartNodeComponent.prototype.node;
    /** @type {?} */
    NgxChartNodeComponent.prototype.hasParent;
    /** @type {?} */
    NgxChartNodeComponent.prototype.direction;
    /** @type {?} */
    NgxChartNodeComponent.prototype.itemClick;
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class NgxChartNodeModule {
}
NgxChartNodeModule.ɵfac = function NgxChartNodeModule_Factory(t) { return new (t || NgxChartNodeModule)(); };
NgxChartNodeModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: NgxChartNodeModule, bootstrap: function () { return [NgxChartNodeComponent]; } });
NgxChartNodeModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[
            CommonModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(NgxChartNodeModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    NgxChartNodeComponent
                ],
                imports: [
                    CommonModule
                ],
                bootstrap: [NgxChartNodeComponent],
                exports: [NgxChartNodeComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NgxChartNodeModule, { declarations: function () { return [NgxChartNodeComponent]; }, imports: function () { return [CommonModule]; }, exports: function () { return [NgxChartNodeComponent]; } }); })();

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class NgxChartDesignerComponent {
    constructor() {
        this.hasParent = false;
        this.direction = 'vertical';
        this.itemClick = new EventEmitter();
    }
    /**
     * @return {?}
     */
    get hostClass() {
        return this.direction === 'vertical' ? 'column' : '';
    }
}
NgxChartDesignerComponent.ɵfac = function NgxChartDesignerComponent_Factory(t) { return new (t || NgxChartDesignerComponent)(); };
NgxChartDesignerComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: NgxChartDesignerComponent, selectors: [["ngx-chart-designer"]], hostVars: 2, hostBindings: function NgxChartDesignerComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵstyleProp("flex-direction", ctx.hostClass);
    } }, inputs: { hasParent: "hasParent", direction: "direction", node: "node" }, outputs: { itemClick: "itemClick" }, decls: 2, vars: 2, consts: [[3, "ngClass", "node", "hasParent", "direction", "itemClick", 4, "ngIf"], ["class", "ngx-org-reports", 3, "ngClass", 4, "ngIf"], [3, "ngClass", "node", "hasParent", "direction", "itemClick"], [1, "ngx-org-reports", 3, "ngClass"], [4, "ngFor", "ngForOf"], [1, "ngx-org-org-container", 3, "ngClass"], [1, "ngx-org-connector-container", 3, "ngClass"], [1, "ngx-org-connector", "ngx-org-border"], [1, "ngx-org-border"], [3, "node", "hasParent", "direction", "itemClick"]], template: function NgxChartDesignerComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, NgxChartDesignerComponent_ngx_chart_node_0_Template, 1, 4, "ngx-chart-node", 0);
        ɵngcc0.ɵɵtemplate(1, NgxChartDesignerComponent_div_1_Template, 2, 2, "div", 1);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", ctx.node);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.node == null ? null : ctx.node.childs == null ? null : ctx.node.childs.length);
    } }, directives: [ɵngcc1.NgIf, NgxChartNodeComponent, ɵngcc1.NgClass, ɵngcc1.NgForOf, NgxChartDesignerComponent], styles: ["[_nghost-%COMP%]{display:flex;align-items:center;flex:1}.ngx-org-vertical[_ngcontent-%COMP%]{flex-direction:column}.ngx-org-org-container[_ngcontent-%COMP%]{display:flex}.ngx-org-org-container-vertical[_ngcontent-%COMP%]{flex-direction:column}.ngx-org-connector[_ngcontent-%COMP%]{flex:1}.ngx-org-connector-container[_ngcontent-%COMP%]{display:flex}.ngx-org-connector-container-horizontal[_ngcontent-%COMP%]{flex-direction:column}.ngx-org-reports[_ngcontent-%COMP%]{display:flex;flex:1}.ngx-org-reports-horizontal[_ngcontent-%COMP%]{flex-direction:column}"] });
NgxChartDesignerComponent.propDecorators = {
    node: [{ type: Input }],
    hasParent: [{ type: Input }],
    direction: [{ type: Input }],
    itemClick: [{ type: Output }],
    hostClass: [{ type: HostBinding, args: ['style.flex-direction',] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(NgxChartDesignerComponent, [{
        type: Component,
        args: [{
                selector: 'ngx-chart-designer',
                template: "<ngx-chart-node *ngIf=\"node\" ngClass=\"ngx-org-{{direction}}\" [node]=\"node\" [hasParent]=\"hasParent\"\r\n  [direction]=\"direction\" (itemClick)=\"itemClick.emit($event)\">\r\n</ngx-chart-node>\r\n\r\n<div *ngIf=\"node?.childs?.length\" ngClass=\"ngx-org-reports-{{direction}}\" class=\"ngx-org-reports\">\r\n  <ng-container *ngFor=\"let childNode of node?.childs; first as isFirst; last as isLast\">\r\n    <div ngClass=\"ngx-org-org-container-{{direction}}\" class=\"ngx-org-org-container\">\r\n      <div ngClass=\"ngx-org-connector-container-{{direction}}\" class=\"ngx-org-connector-container\">\r\n        <div class=\"ngx-org-connector ngx-org-border\" [style.border-color]=\"isFirst?'transparent':''\"></div>\r\n        <div class=\"ngx-org-border\"></div>\r\n        <div class=\"ngx-org-connector ngx-org-border\" [style.border-color]=\"isLast?'transparent':''\"></div>\r\n      </div>\r\n      <ngx-chart-designer [node]=\"childNode\" [hasParent]=\"true\" [direction]=\"direction\" (itemClick)=\"itemClick.emit($event)\">\r\n      </ngx-chart-designer>\r\n    </div>\r\n  </ng-container>\r\n</div>\r\n",
                styles: [":host{display:flex;align-items:center;flex:1}.ngx-org-vertical{flex-direction:column}.ngx-org-org-container{display:flex}.ngx-org-org-container-vertical{flex-direction:column}.ngx-org-connector{flex:1}.ngx-org-connector-container{display:flex}.ngx-org-connector-container-horizontal{flex-direction:column}.ngx-org-reports{display:flex;flex:1}.ngx-org-reports-horizontal{flex-direction:column}"]
            }]
    }], function () { return []; }, { hasParent: [{
            type: Input
        }], direction: [{
            type: Input
        }], itemClick: [{
            type: Output
        }], hostClass: [{
            type: HostBinding,
            args: ['style.flex-direction']
        }], node: [{
            type: Input
        }] }); })();
if (false) {
    /** @type {?} */
    NgxChartDesignerComponent.prototype.node;
    /** @type {?} */
    NgxChartDesignerComponent.prototype.hasParent;
    /** @type {?} */
    NgxChartDesignerComponent.prototype.direction;
    /** @type {?} */
    NgxChartDesignerComponent.prototype.itemClick;
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class NgxChartDesignerModule {
}
NgxChartDesignerModule.ɵfac = function NgxChartDesignerModule_Factory(t) { return new (t || NgxChartDesignerModule)(); };
NgxChartDesignerModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: NgxChartDesignerModule, bootstrap: function () { return [NgxChartDesignerComponent]; } });
NgxChartDesignerModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            NgxChartNodeModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(NgxChartDesignerModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    NgxChartDesignerComponent
                ],
                imports: [
                    CommonModule,
                    NgxChartNodeModule
                ],
                bootstrap: [NgxChartDesignerComponent],
                exports: [NgxChartDesignerComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NgxChartDesignerModule, { declarations: function () { return [NgxChartDesignerComponent]; }, imports: function () { return [CommonModule, NgxChartNodeModule]; }, exports: function () { return [NgxChartDesignerComponent]; } }); })();

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class NgxOrgChartComponent {
    constructor() {
        this.hasParent = false;
        this.direction = 'vertical';
        this.itemClick = new EventEmitter();
    }
}
NgxOrgChartComponent.ɵfac = function NgxOrgChartComponent_Factory(t) { return new (t || NgxOrgChartComponent)(); };
NgxOrgChartComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: NgxOrgChartComponent, selectors: [["ngx-org-chart"]], inputs: { hasParent: "hasParent", direction: "direction", nodes: "nodes" }, outputs: { itemClick: "itemClick" }, decls: 1, vars: 1, consts: [[4, "ngIf"], [4, "ngFor", "ngForOf"], [3, "node", "direction", "itemClick"]], template: function NgxOrgChartComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, NgxOrgChartComponent_div_0_Template, 2, 1, "div", 0);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", ctx.nodes == null ? null : ctx.nodes.length);
    } }, directives: [ɵngcc1.NgIf, ɵngcc1.NgForOf, NgxChartDesignerComponent], styles: ["body[_ngcontent-%COMP%], html[_ngcontent-%COMP%]{display:flex;flex:1}.ngx-org-name[_ngcontent-%COMP%]{font-family:\"Patua One\",cursive}.ngx-org-title[_ngcontent-%COMP%]{font-family:Oswald,sans-serif}.ngx-org-border[_ngcontent-%COMP%]{border-color:#9e9e9e}.ngx-org-box[_ngcontent-%COMP%]{color:#000;width:10em}.ngx-org-self-vertical[_ngcontent-%COMP%]{margin-bottom:2%}"] });
NgxOrgChartComponent.propDecorators = {
    nodes: [{ type: Input }],
    hasParent: [{ type: Input }],
    direction: [{ type: Input }],
    itemClick: [{ type: Output }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(NgxOrgChartComponent, [{
        type: Component,
        args: [{
                selector: 'ngx-org-chart',
                template: "<div *ngIf=\"nodes?.length\">\r\n  <ng-container *ngFor=\"let node of nodes\">\r\n    <div class=\"ngx-org-self-{{direction}}\">\r\n      <ngx-chart-designer [node]=\"node\" [direction]=\"direction\" (itemClick)=\"itemClick.emit($event)\"></ngx-chart-designer>\r\n    </div>\r\n  </ng-container>\r\n</div>\r\n",
                styles: ["body,html{display:flex;flex:1}.ngx-org-name{font-family:\"Patua One\",cursive}.ngx-org-title{font-family:Oswald,sans-serif}.ngx-org-border{border-color:#9e9e9e}.ngx-org-box{color:#000;width:10em}.ngx-org-self-vertical{margin-bottom:2%}"]
            }]
    }], function () { return []; }, { hasParent: [{
            type: Input
        }], direction: [{
            type: Input
        }], itemClick: [{
            type: Output
        }], nodes: [{
            type: Input
        }] }); })();
if (false) {
    /** @type {?} */
    NgxOrgChartComponent.prototype.nodes;
    /** @type {?} */
    NgxOrgChartComponent.prototype.hasParent;
    /** @type {?} */
    NgxOrgChartComponent.prototype.direction;
    /** @type {?} */
    NgxOrgChartComponent.prototype.itemClick;
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class NgxOrgChartModule {
}
NgxOrgChartModule.ɵfac = function NgxOrgChartModule_Factory(t) { return new (t || NgxOrgChartModule)(); };
NgxOrgChartModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: NgxOrgChartModule, bootstrap: function () { return [NgxOrgChartComponent]; } });
NgxOrgChartModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            NgxChartDesignerModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(NgxOrgChartModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    NgxOrgChartComponent
                ],
                imports: [
                    CommonModule,
                    NgxChartDesignerModule
                ],
                bootstrap: [NgxOrgChartComponent],
                exports: [NgxOrgChartComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NgxOrgChartModule, { declarations: function () { return [NgxOrgChartComponent]; }, imports: function () { return [CommonModule, NgxChartDesignerModule]; }, exports: function () { return [NgxOrgChartComponent]; } }); })();

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { NgxOrgChartModule, NgxOrgChartComponent as ɵa, NgxChartDesignerModule as ɵb, NgxChartDesignerComponent as ɵc, NgxChartNodeModule as ɵd, NgxChartNodeComponent as ɵe };
