// be-orgchart/esm2015/lib/clasess/@items/basic/node-basic.js
// getting base64 icons
import { errorBase64Data } from './../../../assets/icons/error.base64';
export class D3NodeBasicParser {
    constructor() {
        // dimensions
        this.width = 240;
        this.height = 100;
        this.frame = {
            type: 'rect',
            className: 'node-rect',
            width: this.width,
            height: this.height,
            stroke: 'blue',
            "stroke-width": 2,
            fill: '#02B2AC',
            rx: 5
        };
        this.defaultImage = {
            type: 'link',
            data: "https://raw.githubusercontent.com/hdcolelia/be-orgchart/master/src/lib/assets/images/root.png"
        };
        // error icon
        this.errorBase64Icon = errorBase64Data;
        this.imageDefs = {
            x: -20, y: -15, h: 60, w: 60, rx: 60
        };
        this.transitionDuration = 600;
    }
    drawNodes(prGroup, prClickEvent) {
        const me = this;
        // adding rect
        me.addRect(prGroup, me.frame);
        //.on('click', (node) => {
        //  prClickEvent.next({ id: node.data.nodeId, node: node.data });
        //})
        // adding title
        me.addText(prGroup, { type: 'text', className: 'title', refField: 'title', x: 50, y: 25, "font-size": 15 });
        me.addText(prGroup, { type: 'text', className: 'description', refField: 'description', x: 50, y: 45, "font-size": 10 });
        me.addText(prGroup, { type: 'text', className: 'description', refField: 'nodeId', x: 50, y: 70, "font-size": 8 });
        // adding image
        const imageUrl = (d) => {
            if (d.data.nodeImage)
                return `url(#img-${d.data.nodeId})`;
            if (me.defaultImage)
                return `url(#img-default-node-img)`;
            return '';
        };
        prGroup.filter(d => imageUrl(d) != '')
            .append('rect').attr('class', 'node-image')
            .attr('x', me.imageDefs.x).attr('y', me.imageDefs.y)
            .attr('width', me.imageDefs.w).attr('height', me.imageDefs.h)
            .attr('fill', d => imageUrl(d)).attr('rx', me.imageDefs.rx);
        // .on('click', (node) => {
        //   prClickEvent.next({ id: node.data.nodeId, node: node.data });
        // });
        // adding click receptor
        prGroup.on('click', (node) => {
            prClickEvent.next({ id: node.data.nodeId, node: node.data });
        });
    }
    //#region adding components
    addRect(gNode, rect) {
        return this.addComponent(gNode, rect);
    }
    addText(gNode, text) {
        return this.addComponent(gNode, text);
    }
    addComponent(gNode, component) {
        const el = gNode.append(component.type);
        el.attr('class', component.className);
        for (const key in component) {
            switch (key) {
                case 'refField':
                    el.text(d => d.data[component[key]]);
                default:
                    el.attr(key, component[key]);
            }
        }
        return el;
    }
}
